IPIP/0000-delegated-routing-http-api.md

```diff
@@ -0,0 +1,88 @@
+and supporting large content providers is a key strategy for driving down IPFS latency.
+These providers must handle high volumes of traffic and support many users, so leveraging industry-standard tools and services
+such as HTTP load balancers, CDNs, reverse proxies, etc. is a requirement.
+To maximize compatibility with standard tools, IPFS needs an HTTP API specification that uses standard HTTP idioms and payload encoding.
+The [Reframe spec](https://github.com/ipfs/specs/blob/main/reframe/REFRAME_PROTOCOL.md) for delegated content routing was an experimental attempt at this, 
+but it has resulted in a very unidiomatic HTTP API which is difficult to implement and is incompatible with many existing tools.
+The cost of a proper redesign, implementation, and maintenance of Reframe and its implementation is too high relative to the urgency of having a delegated routing HTTP API.
+
+Note that this does not supplant nor deprecate Reframe. Ideally in the future, Reframe and its implementation would receive the resources needed to map the IDL to idiomatic HTTP,
+and this spec could then be rewritten in the IDL, maintaining backwards compatibility.
+
+## Detailed design
+
+See the [API design](../routing/DELEGATED_ROUTING_HTTP.md) included with this IPIP.
+
+## Design rationale
+To understand the design rationale, it is important to consider the concrete Reframe limitations that we know about:
+
+- Reframe methods are encoded inside messages
+    - This prevents URL-based pattern matching on methods
+        - Configuring different caching strategies for different methods
+        - Configuring reverse proxies on a per-method basis
+            - Routing methods to specific backends
+            - Method-specific reverse proxy config such as timeouts
+    - Developer UX is poor as a result, e.g. for CDN caching you must encode the entire request message and pass it as a query parameter
+        - This was initially done by URL-escaping the raw bytes
+          - Not possible to consume correctly using standard JavaScript
+          - Shipped in Kubo 0.16
+        - Packing a CID into a struct, encoding it with DAG-CBOR, multibase-encoding that, percent-encoding that, and then passing it in a URL, rather than merely passing the CID in the URL, is needlessly complex from a user's perspective
+        - Added complexity of "Cacheable" methods supporting both POSTs and GETs
+- The required streaming support and message groups add a lot of implementation complexity but isn’t very useful
+    - Ex for FindProviders, the response is buffered anyway for ETag calculation
+    - There are no limits on response sizes nor ways to impose limits and paginate
+    - This is useful for routers that have highly variable resolution time, to send results as soon as possible, but this is not a use case we are focusing on right now and we can add it later
```

Comment:
> It works with at least two routing systems (e.g. DHT and Indexer) -> needs streaming

The reason this doesn't include streaming is because there is no immediate need for it, and we can add it later. The narrow scope is intentional so that we can focus on nailing this particular use case.

"Later" can be mean immediately after this spec; I just don't think it's helpful to block indexer support on it, since they don't need it. For adding it later, consider a streaming response as just a different response format which can be included in content negotiation, such as [ndjson](https://github.com/ndjson/ndjson-spec) with the same provider record schema.

I believe there will remain value in supporting standard `application/json` responses for operators since existing services, infrastructure, middleware, and other tools can produce/parse/manipulate the response without having to write custom code for the format.

> I should be able to make middleware that combines results of various content routing records even if it can't understand some of them

> I should be able to add support for things like HTTP records and BitTorrent records without needing a spec PR. Not every client will support them, but they can if they want to -> extensibility (and not just in the places you know today that you'll need them)

> I should be able to create records related to data retrieval that are "different" in some way from standard records 

What does this mean, concretely? Provider records have a `Protocols` section that can accomodate any multicodec identifier and an opaque JSON value payload, is that not the kind of extensibility you're talking about? Could you provide a complete example that we can work through?

Reply:
From the indexer POV would be great to see paging support so that we can use the protocol with the larger IPFS nodes

Reply:
> The reason this doesn't include streaming is because there is no immediate need for it

I guess it depends on whose use cases you're thinking about, but IMO there are uses for it today (see some support in https://github.com/ipfs/specs/pull/337#discussion_r1005612167). The library https://github.com/libp2p/js-libp2p-delegated-content-routing which is particularly useful in browsers cannot be replaced in a meaningful way with this API, which as I've mentioned means we're not making any progress here other than adding a new API for an existing system (Indexers) which already has an API.

 > "Later" can be mean immediately after this spec; I just don't think it's helpful to block indexer support on it, since they don't need it.

I think it depends. For example, if the v1 protocol doesn't support streaming and v2 does how long will it be expected for the ecosystem to support v1 for? Note that there is already an HTTP API that does exactly what the Indexers want that IMO is largely unsuitable for content routing as a whole (and I suspect most participants in this PR agree or else we'd be pushing to reuse that spec), unless this is really another attempt at the "Indexer HTTP API" with the wrong title name.

>  any multicodec identifier and an opaque JSON value payload

IMO even if we were sticking with this it would still be important for developers to be able to figure out what the data meant which means declaring what they mean here or in a sibling spec.

Reply:
>  For example, if the v1 protocol doesn't support streaming and v2 does how long will it be expected for the ecosystem to support v1 for? 

There is no need for a version bump, it can be introduced in a backwards-compatible way by adding a new content type such as `application/x-ndjson`, and clients and servers can do the usual content negotiation.

> IMO even if we were sticking with this it would still be important for developers to be able to figure out what the data meant which means declaring what they mean here or in a sibling spec.

I agree but I don't think this API should be interpreting the meaning of it, it is pass-through data and asserting semantics here would hinder forwards compatibility. I would hesitate to even call it a "sibling" spec as this API should be completely agnostic to the contents of the payload.

Reply:
> I feel like I've listed multiple examples in my post earlier in this thread which are not supported, but I can work through them more explicitly.

Yeah I get the theoretical ideas, but I am having trouble understanding what it means *concretely*. Are you saying that the peer ID, multiaddr etc. are details of the transfer protocol and should really be part of the opaque payload??  If so, that would make sense and I can change provider records to use tagged unions. What about something like this:

Provider record schema:
```json
{
  "Protocol": "<multicodec_code>",
  ...
}
```

Bitswap provider record (multicodec_code=2320):
```json
{
  "Protocol": "2320",
  "PeerID": "12D3K...",
  "Multiaddrs": ["/ip4...", ...]
} 
```

So then the full response becomes something like

```json
{
  "Providers": [
    {
      "Protocol": "2320",
      "PeerID": "12D3K...",
      "Multiaddrs": ["/ip4...", ...]
    },
    {
      "Protocol": "99999",
      ...
    }
  ]
}
```

Is that like what you have in mind?  (Protocols are stringified multicodec codes so that this is compatible with OpenAPI discriminators, which require discriminator properties to be string values, and they are codes instead of multicodec names because [apparently names aren't as stable as codes](https://github.com/multiformats/multicodec/pull/283).)


> As an example: I should be able to create an endpoint (e.g. at routing.delegate.ipfs.io) that ingests FindProviders responses from say cid.contact, the DHT, and experimental-new-system.tld and responds with both of them as they arrive. If either system adds new types of responses (e.g. new provider types) that shouldn't require updating the proxy in the middle to understand them in order for the new provider types to propagate their way through.

Agreed, that is a design goal of this proposal.